pkg/server/plugin/keymanager/awskms/awskms.go

```diff
@@ -0,0 +1,607 @@
+	}
+
+	keyType, ok := keyTypeFromKeySpec(describeResp.KeyMetadata.CustomerMasterKeySpec)
+	if !ok {
+		return nil, newErrorf(codes.Internal, "unsupported key spec: %v", describeResp.KeyMetadata.CustomerMasterKeySpec)
+	}
+
+	getPublicKeyResp, err := p.kmsClient.GetPublicKey(ctx, &kms.GetPublicKeyInput{KeyId: &aliasName})
+	if err != nil {
+		return nil, newErrorf(codes.Internal, "failed to get public key: %v", err)
+	}
+
+	return &keyEntry{
+		KMSKeyID: *describeResp.KeyMetadata.KeyId,
+		Alias:    aliasName,
+		PublicKey: &keymanager.PublicKey{
+			Id:       spireKeyID,
+			Type:     keyType,
+			PkixData: getPublicKeyResp.PublicKey,
+		},
+	}, err
+}
+
+func (p *Plugin) fetchAliasesPages(ctx context.Context) ([]*keyEntry, error) {
+	var keyEntries []*keyEntry
+	keyEntriesMutex := &sync.Mutex{}
+	paginator := kms.NewListAliasesPaginator(p.kmsClient, &kms.ListAliasesInput{})
+	g, ctx := errgroup.WithContext(ctx)
+
+	for {
+		aliasesResp, err := paginator.NextPage(ctx)
+		if err != nil {
+			return nil, newErrorf(codes.Internal, "failed to build cache of KMS keys: failed to fetch keys: %v", err)
+		}
+		p.log.Debug("Found aliases", "num_aliases", len(aliasesResp.Aliases))
+
+		for _, alias := range aliasesResp.Aliases {
+			spireKeyID, ok := p.spireKeyIDFromAlias(alias)
+			if !ok {
+				continue
+			}
+			if alias.TargetKeyId == nil {
+				// this means something external to the plugin created the alias, without associating it to a key.
+				// it should never happen with CMKs.
+				return nil, newErrorf(codes.FailedPrecondition, "failed to build cache of KMS keys: found SPIRE alias without key: %q", *alias.AliasName)
+			}
+
+			aliasName := *alias.AliasName
+			g.Go(func() error {
+				entry, err := p.fetchKeyEntry(ctx, aliasName, spireKeyID)
+				if err != nil {
+					return err
+				}
+				keyEntriesMutex.Lock()
+				keyEntries = append(keyEntries, entry)
+				keyEntriesMutex.Unlock()
+				return nil
+			})
+		}
+
+		if !paginator.HasMorePages() {
+			break
+		}
+	}
+
+	if err := g.Wait(); err != nil {
+		statusErr := status.Convert(err)
+		return nil, newErrorf(statusErr.Code(), "failed to build cache of KMS keys: %v", statusErr.Message())
+	}
+
+	return keyEntries, nil
+}
+
+// buildCache gets all aliases from KMS and builds an in memory representation with the ones that belong to the current server
+// for each key that belongs to the server we will trigger a gouroutine that gets the extra information
+func (p *Plugin) buildCache(ctx context.Context) error {
+	keyEntries, err := p.fetchAliasesPages(ctx)
+	if err != nil {
+		return err
+	}
+
+	//add results to cache
+	for _, e := range keyEntries {
+		err := p.setEntry(e.PublicKey.Id, *e)
+		if err != nil {
+			return newError(codes.Internal, err.Error())
+		}
+		p.log.Debug("Key loaded", keyIDTag, e.KMSKeyID, aliasTag, e.Alias)
+	}
+
+	return nil
+}
+
+// scheduleDeleteTask ia a long running task that deletes keys that were rotated
```

Comment:
If a KMS key(s) is in use by a set of auto-scale servers, and those servers scale down... or if one of them fails and is re-deployed... I think it is likely that the replacement will have a different prefix configured. Will we leak KMS keys in this case?

Reply:
We will :/

It's "easy" to dispose keys as we rotate them but it's really hard to identify cases (and act on them) as the one you just described.

I don't think it is desirable to delete keys every time a server gracefully shutdowns. I'm some cases we will be right while in others we won't. 

I haven't been able to come up with a better idea when it comes to keys disposal. I wanted to create keys with an expiration date tied to the TTL but the KMS API doesn't seem to support that.

Reply:
@evan2645 @amartinezfayo  @azdagron @rturner3 

A possible way of dealing with scaling up and down could be the next. 

Every time a server starts and fetches all keys, it will also inspect the ones belonging to other Servers in the same cluster.  Making use of the `ca_ttl` from the Server configuration  and `LastUpdatedDate` from the key alias, it is possible to identify the ones that are beyond their time to live and dispose them.

In order to implement this we will need to:

- add the CA TTL to GlobalConfig.
- a way of knowing which cluster each key and Server belong too. 

For the last point we could introduce another prefix, `cluster_prefix`. This will end up building aliases the will look like this `alias/CLUSTER_PREFIX/KEY_PREfIX/KEYMANAGER_KEY_ID`
If we end up introducing a cluster prefix, we should rename `key_prefix` to `server_prefix`.

An alternative is to still use `key_prefix` but instruct users to use a format that allow us to separate the prefix into cluster and server.  Like `key_prefix = CLUSTER_ID/SERVER_ID` or `key_prefix = CLUSTER_ID|SERVER_ID`. This other approach seems more confusing and error prone to me.

Let me know what you think.

Reply:
We could also use the trust domain as a cluster identifier, and build the alias using it.